### ISSUE TITLE
Obtaining a 2D gaussian uncertainty region from input coordinates

### ISSUE THREAD

**Issue by mseglar** [astro-transients/tilepy#15]

**Comment by ednaruiz** [comment]
Hi, I posted this in the LST slack but realized this is the proper place to drop it:

I started to play a bit with implementing the gaussian tiling and realised I was not very far from something usable. The stuff is in a forked version of tilepy and it works well at least tested against the standard GBM tiling. 

https://github.com/ednaruiz/tilepy/blob/gaussian_gbm/src/tilepy/include/MapManagement/MapReader.py#L115 and the test:
https://github.com/ednaruiz/tilepy/blob/gaussian_gbm/examples/launcher/Tiling_Observations_GBM.ipynb

Since the approach is quite modulable, I can also implement the conversion of the _earlier_ localization map sent by GBM glg_locprob_all_*.fit to healpix into the MapReader 
Since I am not familiar with more beyond this in tilepy, let me know if you see another part where this might break and needs further testing.

**Comment by mseglar** [comment]
Hi Edna, I went through the changes and they look good to me! Thanks a lot!

My two comments are that: 
- You can remove the underscore from the function as in python seems to have no effect in Python (as I understand is used for privacy of the function to the class). 
- Those attributes you included make the job but since you used this attribute flexibility,  I am scared now that it is too flexible. The attributes in the function you created are the same, but if for whatever reason dec is changed to decli as input, it wont work (and the attributeError wont be shown in the definition). Also,  very problematic if coordinates are given as lat-lon to the class => that would break the code as lat lon is what is used for the observatory position. To solve it we can add a source to the name, and then in the class ObservationParameter the  __slots__ = ['raSource', 'decSource', 'sigmaSource']. 

 @mdebony had an idea of structuring the MapReader in a certain way, so he might have further structural comments.

**Comment by mseglar** [comment]
for the glg_locprob_all_*.fit to healpix into the MapReader, please go ahead :D I started working on that but I think there is some rotation to apply to obtain the correct healpix projection

**Comment by ednaruiz** [comment]
Thanks @mseglar for the input!
I've put your suggestions, ie, removed the _ for the functions and introduced explicit `raSource`, `decSource`, `sigmaSource` attributes in the `ObservationParameters` class. I also added a helper function to the `MapReader` "`validate_source_params(obspar)`" to raise an error if one of the attributes is missing so that the function does not fail silently. 
A small additional is `set_gaussian_source`, which is defined just before ObservationParameters as a helper function to set the "extra" parameters for the Gaussian case. This is useful to simplify the setup of this case to the user, but not needed:

```
    >>> obspar = ObservationParameters()
    >>> set_gaussian_source(obspar, ra=180.0, dec=30.0, sigma=2.5)
    >>> print(obspar.mode)
    gaussian
```

and the glg_locprob implementation is also there! Also tested with results shown in the notebook. 

Let's hear @mdebony's input on whether the current structure aligns with what is planned for `MapReader`.

**Comment by ednaruiz** [comment]
yes, that is a better way to have the class defined. 
If we agree we can do:

MapReader (as an ABC class)
├── GaussianMapReader
├── LocProbMapReader
└── HealpixMapReader

then instead of calling MapReader(obspars)  we will need a "creator" function create_map_reader(obspar) that has the selection of one of the specific subclasses. Keeping the call of MapReader(obspars) and doing the subclass selection is generally not recommended in python.  Agreed?

**Comment by mdebony** [comment]
> Keeping the call of MapReader(obspars) and doing the subclass selection is generally not recommended in python.
I'm not 100% sure to understand what you meant

**Comment by mseglar** [comment]
I also didnt get that. Another point is that I am going through the code and I see something that is not symmetric to all cases, which is the mode. The last improvements of the map reader went in the opposite direction. In the past, to identify how to read the map, we gave the -type as an arg. We might want to either remove the mode arg (and sort things out in an internal way to the class) or have the mode in all cases.

**Comment by ednaruiz** [comment]
currnelty we do 
`reader = MapReader(obspar)`

with the subclasses implementation doing something like:

`class MapReader:
    def __new__(cls, obspar):
        if obspar.mode == "gaussian":
            return GaussianMapReader(obspar)  # ❌ this causes confusing behavior
        ...`

is poor style. Will work, but for example `isinstance(obj, MapReader)` breaks as we are not returning the class, but the subclass.
So the recommended way is to have subclasses and the MapReader as abstract base class is to implement a creator function 
` reader = create_map_reader(obspar)`
that has all the selection logic inside
`def create_map_reader(obspar) -> MapReader:
    if obspar.mode == "gaussian":
        return GaussianMapReader(obspar)
    elif "glg_locprob_all" in os.path.basename(obspar.skymap):
        return LocProbMapReader(obspar)
    else:
        return HealpixMapReader(obspar)
`

**Comment by mdebony** [comment]
Seem fine to me.

**Comment by mdebony** [comment]
> I also didnt get that. Another point is that I am going through the code and I see something that is not symmetric to all cases, which is the mode. The last improvements of the map reader went in the opposite direction. In the past, to identify how to read the map, we gave the -type as an arg. We might want to either remove the mode arg (and sort things out in an internal way to the class) or have the mode in all cases.

I agree on the way to identify case. But I propose we discuss this more on the PR as implementation specific.

**Comment by ednaruiz** [comment]
about the mode not consistent @mseglar :
we can do something in between and have a `hasattr(obspar, "mode")` inside the `create_map_reader(obspar)`. If mode is given, then we use it. else we fallback to the legacy behaviour of guessing the mode. This gives us the best of two worlds.

**Comment by ednaruiz** [comment]
I will make changes with these two propositions and we can later discuss in a PR as Mathieu suggest

**Comment by aleberti** [comment]
@hashkar I think this one can be closed, the gaussian mode is used since almost an year in LST1 for GBM events, and it works nicely.

**Comment by aleberti** [comment]
@fabianschussler this can be closed as it has been implemented long ago
